## Pandas 

In [1]:
import pandas as pd  #import the pandas library to our notebook

In [176]:
pd.set_option('display.max_columns', 20)  #This command sets the display columns in a dataframe to the number specified
pd.set_option('display.max_rows', 20)   #This command sets the display rows in a dataframe to the number specified

In [158]:
survey = pd.read_csv("Survey_Resp.csv") #this read_csv method in pandas is reading a csv file from local to a data frame

In [160]:
survey[0:3]

RESP_ID  SUR_SCORE         MASKED_IP          CRT_DT COUNTRY_CODE  \
0  13071656          8  MTQ5,Dk1,TMy,A==  1/18/2018 8:48           FI   
1  13071811         10  MTM0Mjk4,TE0OA==  1/18/2018 8:59           FR   
2  13071889          9      MzY4,DgzOTU=  1/18/2018 9:05           AE   

          REGION  
0        uusimaa  
1  ile-de-france  
2          dubai

In [159]:
survey.head(10)

RESP_ID  SUR_SCORE         MASKED_IP          CRT_DT COUNTRY_CODE  \
0  13071656          8  MTQ5,Dk1,TMy,A==  1/18/2018 8:48           FI   
1  13071811         10  MTM0Mjk4,TE0OA==  1/18/2018 8:59           FR   
2  13071889          9      MzY4,DgzOTU=  1/18/2018 9:05           AE   
3  13071909         10  MTM4MzU5MjY2Mg==  1/18/2018 9:06           CZ   
4  13071920         10  MzU3ODcz,Tg1OA==  1/18/2018 9:07           IE   
5  13071988         10  Mjk1,jk1ODE4OQ==  1/18/2018 9:11           HU   
6  13072012          8  MTM2ODUxMTE4,g==  1/18/2018 9:12           GB   
7  13072097          7  MzA1ODE0,zYwOQ==  1/18/2018 9:17           IN   
8  13072253         10  MTg0,DEwMzMx,A==  1/18/2018 9:27           GB   
9  13072647         10  MTI5,zcxMTI0Mg==  1/18/2018 9:50           GB   

          REGION  
0        uusimaa  
1  ile-de-france  
2          dubai  
3          praha  
4            NaN  
5       budapest  
6            NaN  
7  uttar pradesh  
8            NaN  
9            NaN

In [5]:
survey.columns   #will get all the column headers available in the dataset

Index(['RESP_ID', 'SUR_SCORE', 'MASKED_IP', 'CRT_DT', 'COUNTRY_CODE',
       'REGION'],
      dtype='object')

In [6]:
headers = ["RESPONSE_ID","SURVEY_SCORE", "MASKED_IP", "SURVEY_DATE", "COUNTRY_CODE", "REGION"]
survey.columns = headers

In [7]:
survey.columns

Index(['RESPONSE_ID', 'SURVEY_SCORE', 'MASKED_IP', 'SURVEY_DATE',
       'COUNTRY_CODE', 'REGION'],
      dtype='object')

In [8]:
survey.dtypes   #This command will get the datatypes for all the columns in the dataset

RESPONSE_ID      int64
SURVEY_SCORE     int64
MASKED_IP       object
SURVEY_DATE     object
COUNTRY_CODE    object
REGION          object
dtype: object

In [9]:
survey.head(10)

RESPONSE_ID  SURVEY_SCORE         MASKED_IP     SURVEY_DATE COUNTRY_CODE  \
0     13071656             8  MTQ5,Dk1,TMy,A==  1/18/2018 8:48           FI   
1     13071811            10  MTM0Mjk4,TE0OA==  1/18/2018 8:59           FR   
2     13071889             9      MzY4,DgzOTU=  1/18/2018 9:05           AE   
3     13071909            10  MTM4MzU5MjY2Mg==  1/18/2018 9:06           CZ   
4     13071920            10  MzU3ODcz,Tg1OA==  1/18/2018 9:07           IE   
5     13071988            10  Mjk1,jk1ODE4OQ==  1/18/2018 9:11           HU   
6     13072012             8  MTM2ODUxMTE4,g==  1/18/2018 9:12           GB   
7     13072097             7  MzA1ODE0,zYwOQ==  1/18/2018 9:17           IN   
8     13072253            10  MTg0,DEwMzMx,A==  1/18/2018 9:27           GB   
9     13072647            10  MTI5,zcxMTI0Mg==  1/18/2018 9:50           GB   

          REGION  
0        uusimaa  
1  ile-de-france  
2          dubai  
3          praha  
4            NaN  
5       budapest  
6            NaN  
7  uttar pradesh  
8            NaN  
9            NaN

In [10]:
survey.isnull() #Identify the missed values and this method will return the a boolean at cell level based on value exists or not

RESPONSE_ID  SURVEY_SCORE  MASKED_IP  SURVEY_DATE  COUNTRY_CODE  REGION
0           False         False      False        False         False   False
1           False         False      False        False         False   False
2           False         False      False        False         False   False
3           False         False      False        False         False   False
4           False         False      False        False         False    True
...           ...           ...        ...          ...           ...     ...
4783        False         False      False        False         False   False
4784        False         False      False        False         False   False
4785        False         False      False        False         False   False
4786        False         False      False        False          True    True
4787        False         False      False        False         False   False

[4788 rows x 6 columns]

In [11]:
surv_miss_val = survey.isnull() #We are creating anew dataframe based on isnull method acting upon the original dataframe

In [12]:
surv_miss_val.columns.values.tolist()

['RESPONSE_ID',
 'SURVEY_SCORE',
 'MASKED_IP',
 'SURVEY_DATE',
 'COUNTRY_CODE',
 'REGION']

In [13]:
surv_miss_val["REGION"].value_counts()  #This method will return the counts for distinct values existing in the column

False    3215
True     1573
Name: REGION, dtype: int64

##### Now let's count the missing values in each column. "True" represents the missing  values and "False" represents the values exists in Dataset

In [14]:
for col_name in surv_miss_val.columns.values.tolist():  # We are taking each column name into this existing variable
    print("Column Name",col_name)                       # Within the loop we will be print each column name
    print(surv_miss_val[col_name].value_counts())        # And he we are passing each column name to the method to get distinct value counts
    print('------------------------------------------------')    #Just a custom line seperator

Column Name RESPONSE_ID
False    4788
Name: RESPONSE_ID, dtype: int64
------------------------------------------------
Column Name SURVEY_SCORE
False    4788
Name: SURVEY_SCORE, dtype: int64
------------------------------------------------
Column Name MASKED_IP
False    4788
Name: MASKED_IP, dtype: int64
------------------------------------------------
Column Name SURVEY_DATE
False    4788
Name: SURVEY_DATE, dtype: int64
------------------------------------------------
Column Name COUNTRY_CODE
False    3994
True      794
Name: COUNTRY_CODE, dtype: int64
------------------------------------------------
Column Name REGION
False    3215
True     1573
Name: REGION, dtype: int64
------------------------------------------------


#### Survey Response Data Set : Missing data findings
Based on the summary above, out of 4788 rows of data, and 2 columns containing missing data:
<ol>
    <li>"COUNTRY_CODE": 794 missing data</li>
    <li>"REGION": 1573 missing data</li>
</ol>

## Data Analysis

In [15]:
survey.columns

Index(['RESPONSE_ID', 'SURVEY_SCORE', 'MASKED_IP', 'SURVEY_DATE',
       'COUNTRY_CODE', 'REGION'],
      dtype='object')

In [16]:
survey[['SURVEY_SCORE','SURVEY_DATE','COUNTRY_CODE','REGION']].describe(include='all')
# We are tryiong to bring the the descriptive stats in the given dataset based on the columns of our interest 
# and we are using describe method from Pandas Library to achieve that

SURVEY_SCORE     SURVEY_DATE COUNTRY_CODE      REGION
count    4788.000000            4788         3994        3215
unique           NaN            4432          101         365
top              NaN  1/9/2018 17:56           US  california
freq             NaN               6         1468         146
mean        8.119674             NaN          NaN         NaN
std         2.435567             NaN          NaN         NaN
min         0.000000             NaN          NaN         NaN
25%         7.000000             NaN          NaN         NaN
50%         9.000000             NaN          NaN         NaN
75%        10.000000             NaN          NaN         NaN
max        10.000000             NaN          NaN         NaN

In [17]:
survey[['COUNTRY_CODE','RESPONSE_ID','REGION']]

COUNTRY_CODE  RESPONSE_ID         REGION
0              FI     13071656        uusimaa
1              FR     13071811  ile-de-france
2              AE     13071889          dubai
3              CZ     13071909          praha
4              IE     13071920            NaN
...           ...          ...            ...
4783           US     14279849        georgia
4784           US     14280186        florida
4785           US     14280278           ohio
4786          NaN     14281021            NaN
4787           US     14281427          texas

[4788 rows x 3 columns]

In [27]:
a = survey.sort_values(by=['COUNTRY_CODE'],ascending=False, na_position='last')
# The result of the sorting operation is being passed into a dataframe object 'a'


In [31]:
survey.sort_values(by=['COUNTRY_CODE'],ascending=False, inplace=True, na_position='last')
# this will  notsave the results of the sorting operation to the actual dataframe "Survey"

In [32]:
survey.head(10)

RESPONSE_ID  SURVEY_SCORE         MASKED_IP      SURVEY_DATE  \
1214     13010614            10  Mjc2MTk1Mjk0OQ==   1/16/2018 9:39   
2542     13433279             9      ,jk4,jYzODkw  1/31/2018 11:45   
351      13167333             9  Mjgz,TQxOTU0,g==  1/22/2018 10:06   
1161     12987247            10  MzMwMjQzMjkzOQ==  1/15/2018 14:59   
1946     13257960             8      ,jg4,zI3MTMw   1/25/2018 8:45   
1360     13045031            10  MzMx,zE4MTQyMA==  1/17/2018 12:18   
646      12840351            10  Mjgz,TM4MjQ2,A==   1/10/2018 5:59   
359      13168301             6      ,jg5Mjc0OTg2  1/22/2018 10:58   
1447     13071463            10  MzMwMjExODI4OQ==   1/18/2018 8:36   
119      13101159            10  MjY5OTIwMjgxOA==   1/19/2018 8:37   

     COUNTRY_CODE         REGION  
1214           ZM         lusaka  
2542           ZA   western cape  
351            ZA            NaN  
1161           ZA   western cape  
1946           ZA            NaN  
1360           ZA            NaN  
646            ZA            NaN  
359            ZA            NaN  
1447           ZA            NaN  
119            ZA  kwazulu-natal

In [ ]:
survey.sort_values(by=['COUNTRY_CODE'],ascending=False, inplace=True, na_position='last')

In [23]:
a.head(10)

RESPONSE_ID  SURVEY_SCORE         MASKED_IP      SURVEY_DATE  \
1214     13010614            10  Mjc2MTk1Mjk0OQ==   1/16/2018 9:39   
3606     13843072             9      ,jg3OTEzOTAw  2/14/2018 13:27   
3593     13837512             9  MTc2MzQ2,zk0MA==   2/14/2018 9:22   
4256     14082532             9      ,jk3,TAw,TE4  2/22/2018 11:48   
675      12843137            10  MTc3ODM3,zA0Mw==   1/10/2018 9:22   
700      12844909            10  MzMyMDE0,jMz,g==  1/10/2018 11:02   
3273     14337959            10  MTcy,zU4MzI2,w==   3/2/2018 10:40   
1679     13693382            10      ,jk2OTMz,zk1   2/9/2018 11:27   
2900     13566543            10  MjYw,jU5OTU2OQ==   2/5/2018 10:38   
2455     13405934             4      ,jk4,jIwMzc4  1/30/2018 14:37   

     COUNTRY_CODE        REGION  
1214           ZM        lusaka  
3606           ZA           NaN  
3593           ZA           NaN  
4256           ZA           NaN  
675            ZA  western cape  
700            ZA  western cape  
3273           ZA           NaN  
1679           ZA           NaN  
2900           ZA           NaN  
2455           ZA  western cape

In [20]:
a.to_csv('surv_sorted.csv')   # The syntax is <dataframe>.to_csv('path/filename.csv')

In [21]:
b=pd.read_csv('surv_sorted.csv')

In [22]:
b.head(10)

Unnamed: 0  RESPONSE_ID  SURVEY_SCORE         MASKED_IP      SURVEY_DATE  \
0        1214     13010614            10  Mjc2MTk1Mjk0OQ==   1/16/2018 9:39   
1        3606     13843072             9      ,jg3OTEzOTAw  2/14/2018 13:27   
2        3593     13837512             9  MTc2MzQ2,zk0MA==   2/14/2018 9:22   
3        4256     14082532             9      ,jk3,TAw,TE4  2/22/2018 11:48   
4         675     12843137            10  MTc3ODM3,zA0Mw==   1/10/2018 9:22   
5         700     12844909            10  MzMyMDE0,jMz,g==  1/10/2018 11:02   
6        3273     14337959            10  MTcy,zU4MzI2,w==   3/2/2018 10:40   
7        1679     13693382            10      ,jk2OTMz,zk1   2/9/2018 11:27   
8        2900     13566543            10  MjYw,jU5OTU2OQ==   2/5/2018 10:38   
9        2455     13405934             4      ,jk4,jIwMzc4  1/30/2018 14:37   

  COUNTRY_CODE        REGION  
0           ZM        lusaka  
1           ZA           NaN  
2           ZA           NaN  
3           ZA           NaN  
4           ZA  western cape  
5           ZA  western cape  
6           ZA           NaN  
7           ZA           NaN  
8           ZA           NaN  
9           ZA  western cape

In [33]:
a.to_json('surv_sorted.json')   # The syntax is <dataframe>.to_json('path/filename.csv')
# This will create a json file from existing dataframe

In [34]:
json_survey = pd.read_json('surv_sorted.json')
#We are trying to read the Json file that we have created in the previous step

In [35]:
json_survey.head(10)

RESPONSE_ID  SURVEY_SCORE         MASKED_IP      SURVEY_DATE  \
1214     13010614            10  Mjc2MTk1Mjk0OQ==   1/16/2018 9:39   
2542     13433279             9      ,jk4,jYzODkw  1/31/2018 11:45   
351      13167333             9  Mjgz,TQxOTU0,g==  1/22/2018 10:06   
1161     12987247            10  MzMwMjQzMjkzOQ==  1/15/2018 14:59   
1946     13257960             8      ,jg4,zI3MTMw   1/25/2018 8:45   
1360     13045031            10  MzMx,zE4MTQyMA==  1/17/2018 12:18   
646      12840351            10  Mjgz,TM4MjQ2,A==   1/10/2018 5:59   
359      13168301             6      ,jg5Mjc0OTg2  1/22/2018 10:58   
1447     13071463            10  MzMwMjExODI4OQ==   1/18/2018 8:36   
119      13101159            10  MjY5OTIwMjgxOA==   1/19/2018 8:37   

     COUNTRY_CODE         REGION  
1214           ZM         lusaka  
2542           ZA   western cape  
351            ZA           None  
1161           ZA   western cape  
1946           ZA           None  
1360           ZA           None  
646            ZA           None  
359            ZA           None  
1447           ZA           None  
119            ZA  kwazulu-natal

In [45]:
d = [[3,4,6,8,9,0,24,6], ['a','b','c','d','e','f']]
pd.DataFrame(d)
# This is how we can convert a two dimesnional list into a dataframe

0  1  2  3  4  5     6    7
0  3  4  6  8  9  0  24.0  6.0
1  a  b  c  d  e  f   NaN  NaN

In [58]:
e={'CA':{'Prod-1':1, 'Prod-2':2,'Prod-3':3},'NAME':{'Prod-1':'HM','Prod-2':'RV','Prod-3':'SG'},'TX':{'Prod-1':5, 'Prod-2':6,'Prod-3':7},\
   'VT':{'Prod-1':8, 'Prod-2':9,'Prod-3':6}}
df1=pd.DataFrame(e)
#This is how we can create a dataframe from a dictionary
#DataFrame is a predefined class in Pandass library which can convert the Python data structures into a dataframe

In [59]:
df1

CA NAME  TX  VT
Prod-1   1   HM   5   8
Prod-2   2   RV   6   9
Prod-3   3   SG   7   6

In [60]:
df1.mean()
#mean function will return the mean for all the numeric columns in the data frame

CA    2.000000
TX    6.000000
VT    7.666667
dtype: float64

In [61]:
df1.corr()
# Correlation is a process in which we can find and establish a relationship between two or more measures.
# Measures: Measures are the columns which will hold numeric values.
# Dimensions: Dimesions are the columns which will hold non-numeric or string values.

CA        TX        VT
CA  1.000000  1.000000 -0.654654
TX  1.000000  1.000000 -0.654654
VT -0.654654 -0.654654  1.000000

In [63]:
df1.count()

CA      3
NAME    3
TX      3
VT      3
dtype: int64

In [64]:
df1.max()

CA       3
NAME    SG
TX       7
VT       9
dtype: object

In [66]:
df1.min()

CA       1
NAME    HM
TX       5
VT       6
dtype: object

In [67]:
df1.median()

CA    2.0
TX    6.0
VT    8.0
dtype: float64

In [69]:
df1.std()

CA    1.000000
TX    1.000000
VT    1.527525
dtype: float64

In [96]:
df1

CA NAME  TX  VT
Prod-1   1   HM   5   8
Prod-2   2   RV   6   9
Prod-3   3   SG   7   6

In [101]:
df1.iloc[2,3]

6

In [118]:
df1.iloc[-3:-2]

CA NAME  TX  VT
Prod-1   1   HM   5   8

In [107]:
df1.iloc[0:1]

CA NAME  TX  VT
Prod-1   1   HM   5   8

In [113]:
v='BABA'
v[1:]

'AB'

In [106]:
df1.iloc[1:3]

CA NAME  TX  VT
Prod-2   2   RV   6   9
Prod-3   3   SG   7   6

In [135]:
a.columns
#This will print all the columns in the dataframe

Index(['RESPONSE_ID', 'SURVEY_SCORE', 'MASKED_IP', 'SURVEY_DATE',
       'COUNTRY_CODE', 'REGION'],
      dtype='object')

In [166]:
b= a[['SURVEY_SCORE','REGION']].groupby(['REGION']).count()
# This group by operation will return an object which will groupby based on a column 
# and that column will in turn become an index of resultant datafarme 
b.head()

SURVEY_SCORE
REGION                 
aargau                2
abu dhabi             3
aichi                 1
akershus              1
al asimah             1

In [167]:
b.index
# Index method will return the index of a given dataframe

Index(['aargau', 'abu dhabi', 'aichi', 'akershus', 'al asimah',
       'al iskandariyah', 'al kuwayt', 'al qahirah', 'alabama', 'alajuela',
       ...
       'wilayah persekutuan kuala lump', 'wisconsin', 'wyoming', 'yerevan',
       'yucatan', 'zeeland', 'zlinsky kraj', 'zug', 'zuid-holland', 'zurich'],
      dtype='object', name='REGION', length=365)

In [168]:
b.head(3)

SURVEY_SCORE
REGION                 
aargau                2
abu dhabi             3
aichi                 1

In [169]:
b['REGION_CREATED_FROM_INDEX']=b.index
# In the above comman we are trying to copy the index values into a column which we are trying to add to the dataframe
b.head(3)

SURVEY_SCORE REGION_CREATED_FROM_INDEX
REGION                                           
aargau                2                    aargau
abu dhabi             3                 abu dhabi
aichi                 1                     aichi

In [177]:
b.reset_index()

REGION  SURVEY_SCORE REGION_CREATED_FROM_INDEX
0          aargau             2                    aargau
1       abu dhabi             3                 abu dhabi
2           aichi             1                     aichi
3        akershus             1                  akershus
4       al asimah             1                 al asimah
..            ...           ...                       ...
360       zeeland             1                   zeeland
361  zlinsky kraj             1              zlinsky kraj
362           zug             1                       zug
363  zuid-holland            14              zuid-holland
364        zurich            15                    zurich

[365 rows x 3 columns]

In [181]:
b.reset_index(drop=True)
b.head(5)

SURVEY_SCORE REGION_CREATED_FROM_INDEX
REGION                                           
aargau                2                    aargau
abu dhabi             3                 abu dhabi
aichi                 1                     aichi
akershus              1                  akershus
al asimah             1                 al asimah

In [182]:
b.reset_index(drop=True,inplace=True)
b.head(5)

SURVEY_SCORE REGION_CREATED_FROM_INDEX
0             2                    aargau
1             3                 abu dhabi
2             1                     aichi
3             1                  akershus
4             1                 al asimah

In [150]:
b.reset_index(drop=True,inplace=True)
#We are trying to drop and recreate the index based on pandas default indexing
b.head(3)

SURVEY_SCORE REGION_CREATED_FROM_INDEX
0             2                    aargau
1             3                 abu dhabi
2             1                     aichi

In [144]:
headers1 = ['SURVEY_COUNT', 'REGION']
# Create the new column headers we needed to replace in the data frame
b.columns = headers1
# Once we have the headers defined we can pass the headers ito the columns of the dataframe.
b.head(3)

SURVEY_COUNT     REGION
0             2     aargau
1             3  abu dhabi
2             1      aichi